dotnet/src/SemanticKernel.Abstractions/Kernel.cs

```diff
@@ -279,36 +279,32 @@ public CultureInfo Culture
     #region Internal Filtering
 
     [Experimental("SKEXP0001")]
-    internal FunctionInvokingContext? OnFunctionInvokingFilter(KernelFunction function, KernelArguments arguments)
+    internal async Task<FunctionInvocationContext> OnFunctionInvocationAsync(
+        KernelFunction function,
+        KernelArguments arguments,
+        Func<FunctionInvocationContext, Task> functionCallback)
     {
-        FunctionInvokingContext? context = null;
+        FunctionInvocationContext context = new(function, arguments);
 
         if (this._functionFilters is { Count: > 0 })
         {
-            context = new(function, arguments);
-
-            for (int i = 0; i < this._functionFilters.Count; i++)
+            async Task InvokeFilterOrFunctionAsync(FunctionInvocationContext context, int index = 0)
             {
-                this._functionFilters[i].OnFunctionInvoking(context);
+                if (index < this._functionFilters.Count)
+                {
+                    await this._functionFilters[index].OnFunctionInvocationAsync(context, (context) => InvokeFilterOrFunctionAsync(context, index + 1)).ConfigureAwait(false);
+                }
+                else
+                {
+                    await functionCallback(context).ConfigureAwait(false);
+                }
```

Comment:
This code is non trivial so comments would help

Reply:
I simplified the code, moved it to separate method and added the comments.